src/cpp/src/speculative_decoding/speculative_decoding_impl.cpp

```diff
@@ -371,4 +371,516 @@ std::vector<SequenceGroup::Ptr> ContinuousBatchingPipeline::SpeculativeDecodingI
+        OPENVINO_ASSERT(sampling_params[i - 1].adapters == sampling_params[i].adapters,
+                        "LoRA adapters value must be the same for all requests");
+    }
+    auto m_main_eagle_pipeline = std::dynamic_pointer_cast<ContinuousBatchingForEagleDecodingImpl>(m_main_pipeline);
+    auto m_draft_eagle_pipeline = std::dynamic_pointer_cast<ContinuousBatchingForEagleDecodingImpl>(m_draft_pipeline);
+
+    m_main_eagle_pipeline->set_adapters(sampling_params[0].adapters);
+    m_draft_eagle_pipeline->set_adapters(sampling_params[0].adapters);
+    update_eagle_pipeline_params();
+
+    const auto streamer_ptr = std::make_shared<ThreadedStreamerWrapper>(streamer, m_tokenizer);
+
+    OPENVINO_ASSERT(
+        !streamer_ptr->has_callback() || input_ids.size() == 1 && (sampling_params[0].is_greedy_decoding() ||
+                                                                   (sampling_params[0].is_multinomial() &&
+                                                                    sampling_params[0].num_return_sequences == 1)),
+        "Currently eagle streaming is possible only with batch size=1 and only for greedy or multinomial decoding");
+
+    std::vector<GenerationHandle> main_generations;
+    ov::Tensor new_input_ids;
+    for (size_t request_id = 0; request_id < input_ids.size(); ++request_id) {
+        auto new_input_ids = input_ids[request_id];
+        OPENVINO_ASSERT(1 == input_ids[request_id].get_shape().at(0), "Use multiple tensors to pass a batch.");
+        auto main_sampling_params = sampling_params[request_id];
+        OPENVINO_ASSERT(
+            main_sampling_params.assistant_confidence_threshold == 0.f,
+            "Eagle3 Speculative Decoding pipeline only supports `num_assistant_tokens` "
+            "as parameter in GenerationConfig and doesn't work with `assistant_confidence_threshold`.\nPlease "
+            "remove its specification or set it to 0.f.");
+        if (main_sampling_params.num_assistant_tokens == 0) {
+            main_sampling_params.num_assistant_tokens = m_main_pipeline->default_num_assistant_tokens;
+        }
+        main_generations.push_back(
+            m_main_pipeline->add_request(request_id, new_input_ids, main_sampling_params));
+
+        auto draft_sampling_params = main_sampling_params;
+        // set the parameters do not stop draft generation without stopping of the same request for main pipeline
+        draft_sampling_params.ignore_eos = true;
+        draft_sampling_params.stop_strings = {};
+
+        // remove first token from input_ids to create draft_input_ids
+        ov::Tensor draft_input_ids = create_draft_input_ids(new_input_ids);
+
+        std::lock_guard<std::mutex> lock(m_draft_generations_mutex);
+        m_draft_generations.insert(
+            {request_id, m_draft_pipeline->add_request(request_id, draft_input_ids, draft_sampling_params)});
+    }
+    auto all_requests = get_awaiting_requests();
```

Comment:
started from here it's eq to sd pipeline, can you reduce code duplicate and extract different code to separate funtion ?

Reply:
yes, applied, please review again